### PR TITLE
Fixed formatting issue caused by tryAddInfoRightLine().

### LINKED ISSUE #3633: Profile detail display does not truncate excess text
Where the website truncates lengthy text and adds an ellipsis, the game client tries to display the entire text without truncating which causes it to overflow.

[Comparison between game client and web](https://imgur.com/a/rGLsv4j)

### PR DESCRIPTION
**Pull request by swstegall**
Issue #3633 demonstrates that formatting issues in the profile will occur if a supplied string in the profile header is too large. After testing, I found that supplying tryAddInfoRightLine() with a string with more than 34 characters will cause formatting to break. To replicate how this is handled on the osu.Web interface, I added a case where any string with over 34 characters supplied to tryAddInfoRightLine() will print a substring of the first 31 characters and an ellipsis to the string before passing it along to infoTextRight.AddText().

Add any details pertaining to developers above the break.

- [ ] Depends on #PR
- Closes #3633

---

Fixed formatting issue in User ProfileHeader with strings with more than 34 characters.

Add a sentence or two describing this change in plain english. This will be displayed on the [changelog](https://osu.ppy.sh/home/changelog). A single screenshot or short gif is also welcomed.

### CODE REVIEW

**Comment by jorolf** [comment]
You don't have to create 3 different pull requests. You should've added changes to the first one.

**Comment by swstegall** [comment]
> You don't have to create 3 different pull requests. You should've added changes to the first one.

That was my mistake. I'm still fairly new to GitHub and I'm sure my excessive pull requests do not look good.

I still do not understand why my build checks are failing. AppVeyor complains about excessive whitespace on a certain line, but that line contains no additional whitespace, hence why I made the improper choice of duplicating my commit with changes to try and rectify this issue.

File: osu.Game/Overlays/Profile/ProfileHeader.cs [hunk 1]
```diff
@@ -462,16 +462,23 @@ private void loadUser()
 
         private void tryAddInfoRightLine(FontAwesome icon, string str, string url = null)
         {
+            const int LocalCharLimit = 31;
```
Comment (morguldir) [inline osu.Game/Overlays/Profile/ProfileHeader.cs]:
```suggestion
            const int LocalCharLimit = 31;
```

**Review by Krystool — DISMISSED** [review]

File: osu.Game/Overlays/Profile/ProfileHeader.cs [hunk 2]
```diff
@@ -462,16 +462,23 @@ private void loadUser()
 
         private void tryAddInfoRightLine(FontAwesome icon, string str, string url = null)
         {
+            const int local_char_limit = 31;
             if (string.IsNullOrEmpty(str)) return;
 
             infoTextRight.AddIcon(icon);
```
Comment (VPeruS) [inline osu.Game/Overlays/Profile/ProfileHeader.cs]:
```suggestion
             infoTextRight.AddIcon(icon);

             var line = " " + str;
             if (line.Length > local_char_limit)
                 line = str.Substring(0, local_char_limit) + "..."
```

**Comment by Aergwyn** [comment]
This seems to be about the good ol' truncating topic? See https://github.com/ppy/osu/pull/2826 for reference where something similar was discussed.

File: osu.Game/Overlays/Profile/ProfileHeader.cs [hunk 3]
```diff
@@ -462,16 +462,23 @@ private void loadUser()
 
         private void tryAddInfoRightLine(FontAwesome icon, string str, string url = null)
         {
+            const int local_char_limit = 31;
```
Comment (VPeruS) [inline osu.Game/Overlays/Profile/ProfileHeader.cs]: Glyphs can have different width, for example size of "i" != "w".

**Comment by swstegall** [comment]
> This seems to be about the good ol' truncating topic? See #2826 for reference where something similar was discussed.

After analyzing the suggestions in that pull request, I still find myself a bit lost on how this should be handled. Are you suggesting passing the truncation off to another method and then including that method inside tryAddInfoRightLine()? Perhaps something like this:
```csharp
private string Truncate(string str, int bound)
{
    if (str.Length > bound)
        return str.Substring(0, bound) + "...");
}
```

And then calling that method inside tryAddInfoRightLine() with a block like this:
```csharp
private void tryAddInfoRightLine(FontAwesome icon, string str, string url = null)
{
    const int local_char_limit = 31;
    if (string.IsNullOrEmpty(str) == true) return;
    
    infoTextRight.AddIcon(icon);

    var line = " " + str;
    if (line.Length > local_char_limit)
        Truncate(line, local_char_limit);
    if(url != null)
        infoTextRight.AddLink(line, url);
    else
        infoTextRight.AddText(line);

    infoTextRight.NewLine();
}
```

**Comment by Aergwyn** [comment]
No. It was more about that a truncate logic should exist at `SpriteText` level so this kind of localised logic that we'd need at many more places isn't necessary.

**Comment by swstegall** [comment]
> No. It was more about that a truncate logic should exist at `SpriteText` level so this kind of localised logic that we'd need at many more places isn't necessary.

I see. I will have to do some more research and testing before attempting to develop a solution that would work on that scale. Thank you for your patience.

**Review by smoogipoo — CHANGES_REQUESTED** [review]
* Truncation should be implemented inside SpriteText itself (osu!framework).
* You changed the line ends of the entire file.

**Comment by Joehuu** [comment]
Can be closed as referenced issue in OP was resolved with new design.
